staging/src/k8s.io/apiserver/pkg/server/openapi/openapi_handler.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"mime"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-openapi/spec"
+	"github.com/golang/protobuf/proto"
+	"github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
+	"gopkg.in/yaml.v2"
+	genericmux "k8s.io/apiserver/pkg/server/mux"
+)
+
+type OpenAPIService struct {
+	orgSpec      *spec.Swagger
+	specBytes    []byte
+	specPb       []byte
+	specPbGz     []byte
+	lastModified time.Time
+	updateHooks  []func(*http.Request)
+}
+
+// RegisterOpenAPIService registers a handler to provides standard OpenAPI specification.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
+	if !strings.HasSuffix(servePath, JSON_EXT) {
+		return nil, fmt.Errorf("Serving path must ends with \"%s\".", JSON_EXT)
+	}
+
+	servePathBase := servePath[:len(servePath)-len(JSON_EXT)]
+
+	o := OpenAPIService{}
+	if err := o.UpdateSpec(openapiSpec); err != nil {
+		return nil, err
+	}
+
+	mime.AddExtensionType(".json", MIME_JSON)
+	mime.AddExtensionType(".pb-v1", MIME_PB)
+	mime.AddExtensionType(".gz", MIME_PB_GZ)
+
+	type fileInfo struct {
+		ext     string
+		getData func() []byte
+	}
+
+	files := []fileInfo{
+		{".json", o.getSwaggerBytes},
+		{"-2.0.0.json", o.getSwaggerBytes},
+		{"-2.0.0.pb-v1", o.getSwaggerPbBytes},
+		{"-2.0.0.pb-v1.gz", o.getSwaggerPbGzBytes},
+	}
+
+	for _, file := range files {
+		path := servePathBase + file.ext
+		getData := file.getData
+		mux.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+			if r.URL.Path != path {
+				w.WriteHeader(http.StatusNotFound)
+				w.Write([]byte("Path not found!"))
+				return
+			}
+			o.update(r)
```

Comment:
I think the ability to update spec is required specially if we want to support CRDs. A server that has static spec can ignore OpenAPIService and just install the handler. This will make more sense when we just extract openapi code in its own repo. I will continue this discussion in the other PR (that I will send).